### PR TITLE
Backup and Restore lab

### PR DESCRIPTION
**Pull request by codesometech**
Issue #: https://github.com/aws-samples/eks-workshop/issues/726

*Description of changes: Created a lab which shows how to backup and restore EKS cluster using Velero.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by fmedery** [comment]
@codesometech  and @brentley, wouldn't it make more sense to add the chapter to intermediate instead of beginner ?

**Comment by codesometech** [comment]
> @codesometech and @brentley, wouldn't it make more sense to add the chapter to intermediate instead of beginner ?

@fmedery and @brentley  I though it was simple enough to be in beginner section. But I see your point. Cluster backup and restore is typically done by cluster admins and mostly a day 2 operation. I am okay to move it to intermediate if you all agree.

**Comment by brentley** [comment]
I'm fine with this move away from beginner.
--------
Brent
--------

**Comment by paavan98pm** [comment]
@codesometech Thanks for the PR. I'll review it this week and get back to you.

**Comment by codesometech** [comment]
> @codesometech Thanks for the PR. I'll review it this week and get back to you.

Thank you @paavan98pm. Build seems to be failing. I couldn't access the logs. Could you please let me know how to access the build logs.

File: content/intermediate/280_backup-and-restore/deploy-app.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+Verify Persistent Volume Claims
+```
+kubectl get pvc -n staging
+```
+
+Access Wordpress service using port forwording
+```
+kubectl port-forward service/wordpress 8080:80 -n staging
+```
+
```
Comment (paavan98pm) [inline content/intermediate/280_backup-and-restore/deploy-app.md:64]: This fails for me on the local browser. Are you able to access it outside Cloud9 environment on local desktop?

File: content/intermediate/280_backup-and-restore/_index.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+---
+title: "Backup and Restore"
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/_index.md]: Changed the title as per suggestion

File: content/intermediate/280_backup-and-restore/_index.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+weight: 280
+draft: false
+tags:
+  - intermediate
+  - operations
+---
+
+# Backup and Restore EKS using Velero
+
+In this module, you will learn how to backup and restore EKS cluster using Velero.
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/_index.md]: Added intro to velero

File: content/intermediate/280_backup-and-restore/prerequisites.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+title: "Create S3 Bucket and IAM Role for Velero"
+weight: 10
+draft: false
+---
+
+#### Create an S3 bucket to backup cluster
+
+Velero uses AWS S3 bucket to backup EKS cluster. Create an S3 bucket:
+
+{{% notice info %}}
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/prerequisites.md:13]: Removed the manual aspect. Added 'date +%s' in addition to $RANDOM just to be safe.

File: content/intermediate/280_backup-and-restore/prerequisites.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+{{% /notice %}}
+
+{{% notice info %}}
+For us-east-1, use the command below to create S3 bucket. Replace <teamhash or uniquename> with the teamhash that was provided you or use a unique string.
+
+```
+aws s3api create-bucket \
+--bucket eksworkshop-backup-<teamhash or uniquename>-$RANDOM \
+--region $AWS_REGION
+```
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/prerequisites.md:36]: Added description explaining the difference between us-east-1 and other regions when it comes to creating s3 buckets using cli

File: content/intermediate/280_backup-and-restore/prerequisites.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+{
+    "Location": "http://eksworkshop-backup-7asfaff-22480.s3.amazonaws.com/"
+}
+```
+
+Set the BUCKET variable with the bucket name (*example: eksworkshop-backup-7asfaff-22480*) you created.
+```
+BUCKET=<bucketname>
+```
+
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/prerequisites.md:43]: Fixed

File: content/intermediate/280_backup-and-restore/prerequisites.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        "UserName": "velero",
+        "Status": "Active",
+        "CreateDate": "2020-03-18T22:24:41.576Z",
+        "SecretAccessKey": <AWS_SECRET_ACCESS_KEY>,
+        "AccessKeyId": <AWS_ACCESS_KEY_ID>
+  }
+}
+```
+Create a credentials file (velero-credentials) specfic to velero user in your local directory (~/environment):
+
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/prerequisites.md:127]: Velero installation needs credentials file. I removed the manual aspect of creating this file and added some commands to create this file.

File: content/intermediate/280_backup-and-restore/install_velero.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
+---
+title: "Install Velero"
+weight: 20
+draft: false
+---
+
+#### Install Velero binary
+
+Download the [latest release's](https://github.com/vmware-tanzu/velero/releases/latest) tarball for your client platform (Example: velero-v1.3.1-linux-amd64.tar.gz)
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/install_velero.md]: Unfortunately, there is no link to latest. That's why I gave the link to download so users can download the latest manually, but I see your point. I changed it to wget. Tricky part is that there could a new version in a couple of months.

File: content/intermediate/280_backup-and-restore/install_velero.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+Extract the tarball:
+  ``` 
+  tar -xvf <RELEASE-TARBALL-NAME>.tar.gz -C /tmp
+  ```
+Move the extracted velero binary to /usr/local/bin
+```
+mv /tmp/velero-v1.3.1-linux-amd64/velero /usr/local/bin
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/install_velero.md]: Fixed

File: content/intermediate/280_backup-and-restore/install_velero.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+```
+Verify installation
+```
+velero version
+```
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/install_velero.md:24]: Fixed

File: content/intermediate/280_backup-and-restore/deploy-app.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+#### Deploy Wordpress and MySQL in the staging namespace
+
+Create a kubernetes secret for MySQL DB password
+
+```
+kubectl create secret generic mysql-pass --from-literal=password=velerodemo -n staging
+```
+
+```
+cd ~/environment/backup-restore
```
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/deploy-app.md]: Fixed

File: content/intermediate/280_backup-and-restore/deploy-app.md (hunk 1 above)
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/deploy-app.md:78]: Changed to service type LoadBalancer.

File: content/intermediate/280_backup-and-restore/prerequisites.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (codesometech) [reply content/intermediate/280_backup-and-restore/prerequisites.md:127]: @paavan98pm Does this look good? I see that this conversation is unresolved.

**Comment by paavan98pm** [comment]
@codesometech Thanks Praveen for the updates and this useful PR in the Intermediate chapters. I made updates to some typos and other edits in [this](https://github.com/aws-samples/eks-workshop/pull/731/commits/2bfcf553d2fcb291cf7c524f916409582ce1230b) commit and have merged it. It is live on the site now!

**Comment by codesometech** [comment]
> @codesometech Thanks Praveen for the updates and this useful PR in the Intermediate chapters. I made updates to some typos and other edits in [this](https://github.com/aws-samples/eks-workshop/pull/731/commits/2bfcf553d2fcb291cf7c524f916409582ce1230b) commit and have merged it. It is live on the site now!

@paavan98pm Thanks for the edits.

**Pull request merged**